### PR TITLE
Deprecate component.Host.GetExporters

### PR DESCRIPTION
**Pull request by djaglowski**
See #7370

### CODE REVIEW

**Review by mx-psi — APPROVED** [review]
LGTM, should we also deprecate https://github.com/open-telemetry/opentelemetry-collector/blob/042b8763cc7e25cd5dc8a520887822c10bc774a9/service/host.go#L70 and https://github.com/open-telemetry/opentelemetry-collector/blob/042b8763cc7e25cd5dc8a520887822c10bc774a9/service/graph.go#L318

to avoid further internal usage of these methods?

Also, we may need to add exceptions for existing components on contrib that use this method while we migrate them

File: component/host.go [hunk 1]
```diff
@@ -58,5 +58,6 @@ type Host interface {
 	//
 	// GetExporters can be called by the component anytime after Component.Start() begins and
 	// until Component.Shutdown() ends.
+	// Deprecated: [0.74.0] Use a connector or extension to send data between pipelines.
```
Comment (djaglowski) [inline component/host.go:61]:
```suggestion
	// Deprecated: [0.75.0] Use a connector or extension to send data between pipelines.
```

File: service/host.go [hunk 2]
```diff
@@ -68,6 +68,7 @@ func (host *serviceHost) GetExtensions() map[component.ID]component.Component {
 	return host.serviceExtensions.GetExtensions()
 }
 
+// Deprecated: [0.74.0] Use a connector or extension to send data between pipelines.
```
Comment (djaglowski) [inline service/host.go:71]:
```suggestion
// Deprecated: [0.75.0] Use a connector or extension to send data between pipelines.
```

File: service/internal/graph/graph.go [hunk 3]
```diff
@@ -321,6 +321,7 @@ func (g *Graph) ShutdownAll(ctx context.Context) error {
 	return errs
 }
 
+// Deprecated: [0.74.0] Use a connector or extension to send data between pipelines.
```
Comment (djaglowski) [inline service/internal/graph/graph.go:324]:
```suggestion
// Deprecated: [0.75.0] Use a connector or extension to send data between pipelines.
```

**Comment by dmitryax** [comment]
> There are some usages that cannot be converted I believe, one is in the signalfx receiver. cc @dmitryax

At Splunk, we rely on metadata updates sent directly from receivers to exporters. The same approach is implemented in k8scluster, which we have an issue for https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19741. The effort to replace that mechanism is significant. So maybe we can wait until for alternative implementation before the deprecation?

**Comment by dmitryax** [comment]
Also, we usually resolve all the to-be-deprecated usages in contrib before the actual deprecation in the core. Then, the deprecated API is usually removed on the next release. That way, we avoid adding `//nolint` comments in contrib. This deprecation should definitely stay longer than one version, but I think deprecating it now is too aggressive anyway. I'd like to have at least some solution for the metadata updates first.

**Comment by djaglowski** [comment]
> There are some usages that cannot be converted I believe, one is in the signalfx receiver. cc @dmitryax

https://github.com/open-telemetry/opentelemetry-collector/issues/7370 details all uses in core and contrib. `k8sclusterreciever -> signalfxexporter` is the only case that cannot be solved with a connector, but it can be solved with an extension.

> we usually resolve all the to-be-deprecated usages in contrib before the actual deprecation in the core.

In this case at least, I wonder if by waiting to deprecate we could be inviting additional novel uses that would have to be untangled. If you think it'll be a while before `k8sclusterreciever -> signalfxexporter` can use an extension, perhaps we could deprecate when that is the only remaining case? Then we only have one `//nolint` but in exchange we could discourage misuse.

**Comment by dmitryax** [comment]
> https://github.com/open-telemetry/opentelemetry-collector/issues/7370 details all uses in core and contrib. k8sclusterreciever -> signalfxexporter is the only case that cannot be solved with a connector, but it can be solved with an extension.

We also use that in some receivers in Splunk distribution, but it's the same metadata use case.

> In this case at least, I wonder if by waiting to deprecate we could be inviting additional novel uses that would have to be untangled. If you think it'll be a while before k8sclusterreciever -> signalfxexporter can use an extension, perhaps we could deprecate when that is the only remaining case? Then we only have one //nolint but in exchange we could discourage misuse.

Sounds good to me. Let's migrate at least a couple of components before deprecating. And I'll start planning the migration for https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19741. I'm good with deprecating it earlier, but we need a more descriptive message telling users it'll stay until all the usages in contrib are migrated with a link to https://github.com/open-telemetry/opentelemetry-collector/issues/7370. Currently, it seems like it'll be dropped next release as it's usually done for other deprecations.

**Comment by djaglowski** [comment]
> I'm good with deprecating it earlier, but we need a more descriptive message telling users it'll stay until all the usages in contrib are migrated with a link to #7370. Currently, it seems like it'll be dropped next release as it's usually done for other deprecations.

Good points. I'll update the comment and will continue pushing forward with the other uses.

File: component/host.go [hunk 4]
```diff
@@ -58,5 +58,12 @@ type Host interface {
 	//
 	// GetExporters can be called by the component anytime after Component.Start() begins and
 	// until Component.Shutdown() ends.
+	//
+	// Deprecated: [0.75.0] This function will be removed no earlier than 0.80.0.
+	// Several components in the contrib repository use this function so it cannot be removed
```
Comment (dmitryax) [inline component/host.go:63]:
```suggestion
	// Deprecated: [0.75.0] This function will be removed in the future.
	// Several components in the contrib repository use this function so it cannot be removed
```

**Review by dmitryax — APPROVED** [review]
I think we can merge it. Let's just remove the target version for now?

**Review by dmitryax — APPROVED** [review]
I think we can merge it. Let's just remove the target version for now?

File: component/host.go [hunk 5]
```diff
@@ -58,5 +58,12 @@ type Host interface {
 	//
 	// GetExporters can be called by the component anytime after Component.Start() begins and
 	// until Component.Shutdown() ends.
+	//
+	// Deprecated: [0.75.0] This function will be removed in the future.
```
Comment (djaglowski) [inline component/host.go:62]:
```suggestion
	// Deprecated: [0.79.0] This function will be removed in the future.
```

File: service/host.go [hunk 6]
```diff
@@ -68,6 +68,12 @@ func (host *serviceHost) GetExtensions() map[component.ID]component.Component {
 	return host.serviceExtensions.GetExtensions()
 }
 
+// Deprecated: [0.75.0] This function will be removed no earlier than 0.80.0.
```
Comment (djaglowski) [inline service/host.go:71]:
```suggestion
// Deprecated: [0.79.0] This function will be removed in the future.
```

File: service/internal/graph/graph.go [hunk 7]
```diff
@@ -321,6 +321,12 @@ func (g *Graph) ShutdownAll(ctx context.Context) error {
 	return errs
 }
 
+// Deprecated: [0.75.0] This function will be removed no earlier than 0.80.0.
```
Comment (djaglowski) [inline service/internal/graph/graph.go:324]:
```suggestion
// Deprecated: [0.79.0] This function will be removed in the future.
```

**Review by mx-psi — APPROVED** [review]

**Review by dmitryax — APPROVED** [review]

**Pull request merged**